lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -98,181 +128,457 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+                        allocatedEvaluator.Id);
             }
+            allocatedEvaluator.SubmitContextAndService(configs.Context, configs.Service);
         }
+        #endregion IAllocatedEvaluator
 
+        #region IActiveContext
         /// <summary>
-        /// Specifies what to do when the task is completed
-        /// In this case just disposes off the task
+        /// IActiveContext handler. It will take the following actions based on the system state:
+        /// Case WaitingForEvaluator:
+        ///    Adds Active Context to Active Context Manager
+        /// Case Fail:
+        ///    Closes the ActiveContext
+        /// Other cases - not expected
         /// </summary>
-        /// <param name="completedTask">The link to the completed task</param>
-        public void OnNext(ICompletedTask completedTask)
+        /// <param name="activeContext"></param>
+        public void OnNext(IActiveContext activeContext)
         {
-            lock (_completedTasks)
+            Logger.Log(Level.Verbose, string.Format(CultureInfo.InvariantCulture, "Received Active Context {0}, systemState {1}.", activeContext.Id, _systemState.CurrentState));
+            lock (_lock)
             {
-                Logger.Log(Level.Info,
-                    string.Format("Received completed task message from task Id: {0}", completedTask.Id));
-                _completedTasks.Add(completedTask);
-
-                if (AreIMRUTasksCompleted())
+                switch (_systemState.CurrentState)
                 {
-                    ShutDownAllEvaluators();
+                    case SystemState.WaitingForEvaluator:
+                        _contextManager.Add(activeContext);
+                        break;
+                    case SystemState.Fail:
+                        Logger.Log(Level.Info, "Received IActiveContext event, but system is in FAIL state. Closing the context.");
+                        activeContext.Dispose();
+                        break;
+                    default:
+                        UnexpectedState(activeContext.Id, "IActiveContext");
+                        break;
                 }
             }
         }
+        #endregion IActiveContext
 
+        #region submit tasks
         /// <summary>
-        /// Specifies what to do when evaluator fails.
-        /// If we get all completed tasks then ignore the failure
-        /// Else request a new evaluator. If failure happens in middle of IMRU 
-        /// job we expect neighboring evaluators to fail while doing 
-        /// communication and will use FailedTask and FailedContext logic to 
-        /// order shutdown.
+        /// Called from ActiveContextManager when all the expected active context are received.
+        /// It changes the system state then calls SubmitTasks().
         /// </summary>
         /// <param name="value"></param>
-        public void OnNext(IFailedEvaluator value)
+        public void OnNext(int value)
```

Comment:
We can only start to submit tasks after all the the active context are received. This is required by IMRU/GC, it also makes the fault tolerant system state transition clean. Please refer to the design document for details which is attached to REEF-1223. 
We had several discussions about this and have decided to use notification. We don't want to keep checking it in IActiveContext  handler. Instead, we want ContextManager to notify Driver to start submitting tasks once all contexts are received. We can achieve the same purpose by many different ways in coding, but this is what we talked and agreed on. Unless you have strong reason to prefer one way to another, then we can bring for discussion again. 

I changed code to use OnNext( IAllContextsReceived value) as you suggested. It added two additional classes, makes a simple issue little bit more complex, but makes the code easy to read.